symengine/utilities/catch/CMakeLists.txt

```diff
@@ -4,5 +4,5 @@ set(SRC
     catch.cpp
 )
 
-add_library(catch STATIC ${SRC})
+add_library(catch ${SRC})
 target_link_libraries(catch symengine)
```

Comment:
I think the better option would be to not link `symengine` here.

Reply:
With
```
add_library(catch STATIC ${SRC})
#target_link_libraries(catch symengine)
```
I end up with these link issues. I think they're the same as I had before my proposed patch.
```
[ 58%] Linking CXX executable test_basic
ld: warning: ignoring file ../../utilities/catch/libcatch.a, file was built for archive which is not the architecture being linked (x86_64): ../../utilities/catch/libcatch.a
ld: warning: ld: warning: ignoring file ../../utilities/catch/libcatch.a, file was built for archive which is not the architecture being linked (x86_64): ../../utilities/catch/libcatch.aignoring file ../../utilities/catch/libcatch.a, file was built for archive which is not the architecture being linked (x86_64): ../../utilities/catch/libcatch.a

ld: warning: ignoring file ../../utilities/catch/libcatch.a, file was built for archive which is not the architecture being linked (x86_64): ../../utilities/catch/libcatch.a
Undefined symbols for architecture x86_64:
  "Catch::ResultBuilder::captureResult(Catch::ResultWas::OfType)", referenced from:
      ____C_A_T_C_H____T_E_S_T____66() in test_sets.cpp.o
      ____C_A_T_C_H____T_E_S_T____225() in test_sets.cpp.o
      ____C_A_T_C_H____T_E_S_T____254() in test_sets.cpp.o
      ____C_A_T_C_H____T_E_S_T____485() in test_sets.cpp.o
      ____C_A_T_C_H____T_E_S_T____520() in test_sets.cpp.o
      ____C_A_T_C_H____T_E_S_T____741() in test_sets.cpp.o
  "Catch::ResultBuilder::endExpression()", referenced from:
      ____C_A_T_C_H____T_E_S_T____66() in test_sets.cpp.o
      ____C_A_T_C_H____T_E_S_T____225() in test_sets.cpp.o
      ____C_A_T_C_H____T_E_S_T____254() in test_sets.cpp.o
      ____C_A_T_C_H____T_E_S_T____293() in test_sets.cpp.o
      ____C_A_T_C_H____T_E_S_T____403() in test_sets.cpp.o
      ____C_A_T_C_H____T_E_S_T____485() in test_sets.cpp.o
      ____C_A_T_C_H____T_E_S_T____632() in test_sets.cpp.o
```

The tested configure settings are
```
cmake \
-DCMAKE_C_COMPILER="$SPACK_VIEW_DIR/bin/mpicc" \
-DCMAKE_CXX_COMPILER="$SPACK_VIEW_DIR/bin/mpicxx" \
-DCMAKE_BUILD_TYPE=Release \
-DINTEGER_CLASS:STRING=boostmp \
-DBoost_INCLUDE_DIR=$SPACK_VIEW_DIR/include \
-DWITH_LLVM=ON \
-DWITH_COTIRE=OFF \
-DBUILD_SHARED_LIBS:BOOL=on \
-DBUILD_TESTS=TRUE \
-DBUILD_BENCHMARKS=TRUE \
-DCMAKE_INSTALL_RPATH_USE_LINK_PATH:BOOL=on \
-DCMAKE_INSTALL_PREFIX="<path>/SymEngine/install" \
../
```

Reply:
What is `CMAKE_AR` in `CMakeCache.txt`?

Reply:
```
add_library(catch ${SRC})
#target_link_libraries(catch symengine)
```
still works fine for me though, but it appears that the tester doesn't like the change?

Reply:
I think the issue is that the archiver is set to something different, so none of the static libraries can be linked to create an executable. What is `CMAKE_AR` and `CMAKE_CXX_COMPILER` set to?

Reply:
The variable `CMAKE_AR` doesn't exist in my [CMakeCache.txt (attached)](https://github.com/symengine/symengine/files/1483215/CMakeCache.txt). In case its more insightful, here's a full dump of the link line. Thanks for the help BTW!
```
$ make VERBOSE=1
/Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/cmake-3.7.2-dltxeqnisp5b4nt7c6vs3dhgndrju7ct/bin/cmake -H/Users/jp/Programs/GitHub/SymEngine/symengine -B/Users/jp/Programs/GitHub/SymEngine/symengine/build --check-build-system CMakeFiles/Makefile.cmake 0
/Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/cmake-3.7.2-dltxeqnisp5b4nt7c6vs3dhgndrju7ct/bin/cmake -E cmake_progress_start /Users/jp/Programs/GitHub/SymEngine/symengine/build/CMakeFiles /Users/jp/Programs/GitHub/SymEngine/symengine/build/CMakeFiles/progress.marks
/Library/Developer/CommandLineTools/usr/bin/make -f CMakeFiles/Makefile2 all
/Library/Developer/CommandLineTools/usr/bin/make -f symengine/CMakeFiles/symengine.dir/build.make symengine/CMakeFiles/symengine.dir/depend
cd /Users/jp/Programs/GitHub/SymEngine/symengine/build && /Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/cmake-3.7.2-dltxeqnisp5b4nt7c6vs3dhgndrju7ct/bin/cmake -E cmake_depends "Unix Makefiles" /Users/jp/Programs/GitHub/SymEngine/symengine /Users/jp/Programs/GitHub/SymEngine/symengine/symengine /Users/jp/Programs/GitHub/SymEngine/symengine/build /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine/CMakeFiles/symengine.dir/DependInfo.cmake --color=
/Library/Developer/CommandLineTools/usr/bin/make -f symengine/CMakeFiles/symengine.dir/build.make symengine/CMakeFiles/symengine.dir/build
make[2]: Nothing to be done for `symengine/CMakeFiles/symengine.dir/build'.
[ 30%] Built target symengine
/Library/Developer/CommandLineTools/usr/bin/make -f symengine/utilities/catch/CMakeFiles/catch.dir/build.make symengine/utilities/catch/CMakeFiles/catch.dir/depend
cd /Users/jp/Programs/GitHub/SymEngine/symengine/build && /Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/cmake-3.7.2-dltxeqnisp5b4nt7c6vs3dhgndrju7ct/bin/cmake -E cmake_depends "Unix Makefiles" /Users/jp/Programs/GitHub/SymEngine/symengine /Users/jp/Programs/GitHub/SymEngine/symengine/symengine/utilities/catch /Users/jp/Programs/GitHub/SymEngine/symengine/build /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine/utilities/catch /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine/utilities/catch/CMakeFiles/catch.dir/DependInfo.cmake --color=
/Library/Developer/CommandLineTools/usr/bin/make -f symengine/utilities/catch/CMakeFiles/catch.dir/build.make symengine/utilities/catch/CMakeFiles/catch.dir/build
make[2]: Nothing to be done for `symengine/utilities/catch/CMakeFiles/catch.dir/build'.
[ 31%] Built target catch
/Library/Developer/CommandLineTools/usr/bin/make -f symengine/tests/rcp/CMakeFiles/test_rcp.dir/build.make symengine/tests/rcp/CMakeFiles/test_rcp.dir/depend
cd /Users/jp/Programs/GitHub/SymEngine/symengine/build && /Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/cmake-3.7.2-dltxeqnisp5b4nt7c6vs3dhgndrju7ct/bin/cmake -E cmake_depends "Unix Makefiles" /Users/jp/Programs/GitHub/SymEngine/symengine /Users/jp/Programs/GitHub/SymEngine/symengine/symengine/tests/rcp /Users/jp/Programs/GitHub/SymEngine/symengine/build /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine/tests/rcp /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine/tests/rcp/CMakeFiles/test_rcp.dir/DependInfo.cmake --color=
/Library/Developer/CommandLineTools/usr/bin/make -f symengine/tests/rcp/CMakeFiles/test_rcp.dir/build.make symengine/tests/rcp/CMakeFiles/test_rcp.dir/build
[ 32%] Linking CXX executable test_rcp
cd /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine/tests/rcp && /Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/cmake-3.7.2-dltxeqnisp5b4nt7c6vs3dhgndrju7ct/bin/cmake -E cmake_link_script CMakeFiles/test_rcp.dir/link.txt --verbose=1
/Users/jp/Programs/GitHub/Spack/views/deal_II_dev/bin/mpicxx   -std=c++11 -Wall -Wextra -fPIC  -ffast-math -funroll-loops -Wno-unused-parameter -march=native -D__extern_always_inline=inline -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -DNDEBUG -Wno-unknown-pragmas -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/test_rcp.dir/test_rcp.cpp.o  -o test_rcp -Wl,-rpath,/Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine -Wl,-rpath,/Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib  ../../utilities/catch/libcatch.a ../../libsymengine.0.3.0.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMAsmParser.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMCore.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMExecutionEngine.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMInstCombine.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMMCJIT.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMX86CodeGen.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMX86Desc.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMX86Info.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMScalarOpts.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMSupport.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMTransformUtils.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMX86AsmPrinter.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMX86AsmParser.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMX86Disassembler.dylib
ld: warning: ignoring file ../../utilities/catch/libcatch.a, file was built for archive which is not the architecture being linked (x86_64): ../../utilities/catch/libcatch.a
Undefined symbols for architecture x86_64:
  "Catch::ResultBuilder::endExpression()", referenced from:
      ____C_A_T_C_H____T_E_S_T____19() in test_rcp.cpp.o
      f(Mesh&) in test_rcp.cpp.o
      f_const(Mesh const&) in test_rcp.cpp.o
      ____C_A_T_C_H____T_E_S_T____49() in test_rcp.cpp.o
      ____C_A_T_C_H____T_E_S_T____63() in test_rcp.cpp.o
      f2_const(Mesh2 const&) in test_rcp.cpp.o
      f2_hybrid(Mesh2&) in test_rcp.cpp.o
      ...
  "Catch::ResultBuilder::setResultType(bool)", referenced from:
      Catch::ExpressionLhs<bool>::endExpression() in test_rcp.cpp.o
      Catch::ResultBuilder& Catch::ExpressionLhs<int const&>::captureExpression<(Catch::Internal::Operator)0, int>(int const&) in test_rcp.cpp.o
      Catch::ResultBuilder& Catch::ExpressionLhs<unsigned int const&>::captureExpression<(Catch::Internal::Operator)0, int>(int const&) in test_rcp.cpp.o
  "Catch::ResultBuilder::useActiveException(Catch::ResultDisposition::Flags)", referenced from:
      ____C_A_T_C_H____T_E_S_T____19() in test_rcp.cpp.o
      f(Mesh&) in test_rcp.cpp.o
      f_const(Mesh const&) in test_rcp.cpp.o
      ____C_A_T_C_H____T_E_S_T____49() in test_rcp.cpp.o
      ____C_A_T_C_H____T_E_S_T____63() in test_rcp.cpp.o
      f2_const(Mesh2 const&) in test_rcp.cpp.o
      f2_hybrid(Mesh2&) in test_rcp.cpp.o
      ...
  "Catch::ResultBuilder::react()", referenced from:
      ____C_A_T_C_H____T_E_S_T____19() in test_rcp.cpp.o
      f(Mesh&) in test_rcp.cpp.o
      f_const(Mesh const&) in test_rcp.cpp.o
      ____C_A_T_C_H____T_E_S_T____49() in test_rcp.cpp.o
      ____C_A_T_C_H____T_E_S_T____63() in test_rcp.cpp.o
      f2_const(Mesh2 const&) in test_rcp.cpp.o
      f2_hybrid(Mesh2&) in test_rcp.cpp.o
      ...
  "Catch::ResultBuilder::setOp(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      Catch::ResultBuilder& Catch::ExpressionLhs<int const&>::captureExpression<(Catch::Internal::Operator)0, int>(int const&) in test_rcp.cpp.o
      Catch::ResultBuilder& Catch::ExpressionLhs<unsigned int const&>::captureExpression<(Catch::Internal::Operator)0, int>(int const&) in test_rcp.cpp.o
  "Catch::ResultBuilder::setLhs(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      Catch::ExpressionLhs<bool>::endExpression() in test_rcp.cpp.o
      Catch::ResultBuilder& Catch::ExpressionLhs<int const&>::captureExpression<(Catch::Internal::Operator)0, int>(int const&) in test_rcp.cpp.o
      Catch::ResultBuilder& Catch::ExpressionLhs<unsigned int const&>::captureExpression<(Catch::Internal::Operator)0, int>(int const&) in test_rcp.cpp.o
  "Catch::ResultBuilder::setRhs(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)", referenced from:
      Catch::ResultBuilder& Catch::ExpressionLhs<int const&>::captureExpression<(Catch::Internal::Operator)0, int>(int const&) in test_rcp.cpp.o
      Catch::ResultBuilder& Catch::ExpressionLhs<unsigned int const&>::captureExpression<(Catch::Internal::Operator)0, int>(int const&) in test_rcp.cpp.o
  "Catch::ResultBuilder::ResultBuilder(char const*, Catch::SourceLineInfo const&, char const*, Catch::ResultDisposition::Flags)", referenced from:
      ____C_A_T_C_H____T_E_S_T____19() in test_rcp.cpp.o
      f(Mesh&) in test_rcp.cpp.o
      f_const(Mesh const&) in test_rcp.cpp.o
      ____C_A_T_C_H____T_E_S_T____49() in test_rcp.cpp.o
      ____C_A_T_C_H____T_E_S_T____63() in test_rcp.cpp.o
      f2_const(Mesh2 const&) in test_rcp.cpp.o
      f2_hybrid(Mesh2&) in test_rcp.cpp.o
      ...
  "Catch::SourceLineInfo::SourceLineInfo(char const*, unsigned long)", referenced from:
      ____C_A_T_C_H____T_E_S_T____19() in test_rcp.cpp.o
      f(Mesh&) in test_rcp.cpp.o
      f_const(Mesh const&) in test_rcp.cpp.o
      ____C_A_T_C_H____T_E_S_T____49() in test_rcp.cpp.o
      ____C_A_T_C_H____T_E_S_T____63() in test_rcp.cpp.o
      f2_const(Mesh2 const&) in test_rcp.cpp.o
      f2_hybrid(Mesh2&) in test_rcp.cpp.o
      ...
  "Catch::AutoReg::AutoReg(void (*)(), Catch::SourceLineInfo const&, Catch::NameAndDesc const&)", referenced from:
      ___cxx_global_var_init in test_rcp.cpp.o
      ___cxx_global_var_init.6 in test_rcp.cpp.o
      ___cxx_global_var_init.8 in test_rcp.cpp.o
      ___cxx_global_var_init.10 in test_rcp.cpp.o
  "Catch::AutoReg::~AutoReg()", referenced from:
      ___cxx_global_var_init in test_rcp.cpp.o
      ___cxx_global_var_init.6 in test_rcp.cpp.o
      ___cxx_global_var_init.8 in test_rcp.cpp.o
      ___cxx_global_var_init.10 in test_rcp.cpp.o
  "Catch::toString(bool)", referenced from:
      Catch::ExpressionLhs<bool>::endExpression() in test_rcp.cpp.o
  "Catch::toString(int)", referenced from:
      Catch::ResultBuilder& Catch::ExpressionLhs<int const&>::captureExpression<(Catch::Internal::Operator)0, int>(int const&) in test_rcp.cpp.o
      Catch::ResultBuilder& Catch::ExpressionLhs<unsigned int const&>::captureExpression<(Catch::Internal::Operator)0, int>(int const&) in test_rcp.cpp.o
  "Catch::toString(unsigned int)", referenced from:
      Catch::ResultBuilder& Catch::ExpressionLhs<unsigned int const&>::captureExpression<(Catch::Internal::Operator)0, int>(int const&) in test_rcp.cpp.o
  "Catch::ResultBuilder::shouldDebugBreak() const", referenced from:
      ____C_A_T_C_H____T_E_S_T____19() in test_rcp.cpp.o
      f(Mesh&) in test_rcp.cpp.o
      f_const(Mesh const&) in test_rcp.cpp.o
      ____C_A_T_C_H____T_E_S_T____49() in test_rcp.cpp.o
      ____C_A_T_C_H____T_E_S_T____63() in test_rcp.cpp.o
      f2_const(Mesh2 const&) in test_rcp.cpp.o
      f2_hybrid(Mesh2&) in test_rcp.cpp.o
      ...
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [symengine/tests/rcp/test_rcp] Error 1
```

Reply:
Can you delete `../../utilities/catch/libcatch.a` and run `make VERBOSE=1` again?

Reply:
I reverted to ``, nuked the build directory and built again. Still no luck.

Here's how `libcatch.a` is built
```
$ make VERBOSE=1
/Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/cmake-3.7.2-dltxeqnisp5b4nt7c6vs3dhgndrju7ct/bin/cmake -H/Users/jp/Programs/GitHub/SymEngine/symengine -B/Users/jp/Programs/GitHub/SymEngine/symengine/build --check-build-system CMakeFiles/Makefile.cmake 0
/Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/cmake-3.7.2-dltxeqnisp5b4nt7c6vs3dhgndrju7ct/bin/cmake -E cmake_progress_start /Users/jp/Programs/GitHub/SymEngine/symengine/build/CMakeFiles /Users/jp/Programs/GitHub/SymEngine/symengine/build/CMakeFiles/progress.marks
/Library/Developer/CommandLineTools/usr/bin/make -f CMakeFiles/Makefile2 all
/Library/Developer/CommandLineTools/usr/bin/make -f symengine/CMakeFiles/symengine.dir/build.make symengine/CMakeFiles/symengine.dir/depend
cd /Users/jp/Programs/GitHub/SymEngine/symengine/build && /Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/cmake-3.7.2-dltxeqnisp5b4nt7c6vs3dhgndrju7ct/bin/cmake -E cmake_depends "Unix Makefiles" /Users/jp/Programs/GitHub/SymEngine/symengine /Users/jp/Programs/GitHub/SymEngine/symengine/symengine /Users/jp/Programs/GitHub/SymEngine/symengine/build /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine/CMakeFiles/symengine.dir/DependInfo.cmake --color=
/Library/Developer/CommandLineTools/usr/bin/make -f symengine/CMakeFiles/symengine.dir/build.make symengine/CMakeFiles/symengine.dir/build
make[2]: Nothing to be done for `symengine/CMakeFiles/symengine.dir/build'.
[ 30%] Built target symengine
/Library/Developer/CommandLineTools/usr/bin/make -f symengine/utilities/catch/CMakeFiles/catch.dir/build.make symengine/utilities/catch/CMakeFiles/catch.dir/depend
cd /Users/jp/Programs/GitHub/SymEngine/symengine/build && /Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/cmake-3.7.2-dltxeqnisp5b4nt7c6vs3dhgndrju7ct/bin/cmake -E cmake_depends "Unix Makefiles" /Users/jp/Programs/GitHub/SymEngine/symengine /Users/jp/Programs/GitHub/SymEngine/symengine/symengine/utilities/catch /Users/jp/Programs/GitHub/SymEngine/symengine/build /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine/utilities/catch /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine/utilities/catch/CMakeFiles/catch.dir/DependInfo.cmake --color=
/Library/Developer/CommandLineTools/usr/bin/make -f symengine/utilities/catch/CMakeFiles/catch.dir/build.make symengine/utilities/catch/CMakeFiles/catch.dir/build
[ 31%] Linking CXX static library libcatch.a
cd /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine/utilities/catch && /Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/cmake-3.7.2-dltxeqnisp5b4nt7c6vs3dhgndrju7ct/bin/cmake -P CMakeFiles/catch.dir/cmake_clean_target.cmake
cd /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine/utilities/catch && /Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/cmake-3.7.2-dltxeqnisp5b4nt7c6vs3dhgndrju7ct/bin/cmake -E cmake_link_script CMakeFiles/catch.dir/link.txt --verbose=1
/Users/jp/Programs/GitHub/Spack/views/deal_II_dev/bin/ar qc libcatch.a  CMakeFiles/catch.dir/catch.cpp.o
/Users/jp/Programs/GitHub/Spack/views/deal_II_dev/bin/ranlib libcatch.a
[ 31%] Built target catch
```
and the failing test (again)
```
/Library/Developer/CommandLineTools/usr/bin/make -f symengine/tests/rcp/CMakeFiles/test_rcp.dir/build.make symengine/tests/rcp/CMakeFiles/test_rcp.dir/depend
cd /Users/jp/Programs/GitHub/SymEngine/symengine/build && /Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/cmake-3.7.2-dltxeqnisp5b4nt7c6vs3dhgndrju7ct/bin/cmake -E cmake_depends "Unix Makefiles" /Users/jp/Programs/GitHub/SymEngine/symengine /Users/jp/Programs/GitHub/SymEngine/symengine/symengine/tests/rcp /Users/jp/Programs/GitHub/SymEngine/symengine/build /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine/tests/rcp /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine/tests/rcp/CMakeFiles/test_rcp.dir/DependInfo.cmake --color=
/Library/Developer/CommandLineTools/usr/bin/make -f symengine/tests/rcp/CMakeFiles/test_rcp.dir/build.make symengine/tests/rcp/CMakeFiles/test_rcp.dir/build
[ 32%] Linking CXX executable test_rcp
cd /Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine/tests/rcp && /Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/cmake-3.7.2-dltxeqnisp5b4nt7c6vs3dhgndrju7ct/bin/cmake -E cmake_link_script CMakeFiles/test_rcp.dir/link.txt --verbose=1
/Users/jp/Programs/GitHub/Spack/views/deal_II_dev/bin/mpicxx   -std=c++11 -Wall -Wextra -fPIC  -ffast-math -funroll-loops -Wno-unused-parameter -march=native -D__extern_always_inline=inline -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -DNDEBUG -Wno-unknown-pragmas -Wl,-search_paths_first -Wl,-headerpad_max_install_names  CMakeFiles/test_rcp.dir/test_rcp.cpp.o  -o test_rcp -Wl,-rpath,/Users/jp/Programs/GitHub/SymEngine/symengine/build/symengine -Wl,-rpath,/Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib  ../../utilities/catch/libcatch.a ../../libsymengine.0.3.0.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMAsmParser.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMCore.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMExecutionEngine.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMInstCombine.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMMCJIT.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMX86CodeGen.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMX86Desc.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMX86Info.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMScalarOpts.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMSupport.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMTransformUtils.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMX86AsmPrinter.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMX86AsmParser.dylib /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/lib/libLLVMX86Disassembler.dylib
ld: warning: ignoring file ../../utilities/catch/libcatch.a, file was built for archive which is not the architecture being linked (x86_64): ../../utilities/catch/libcatch.a
Undefined symbols for architecture x86_64:
  "Catch::ResultBuilder::endExpression()", referenced from:
      ____C_A_T_C_H____T_E_S_T____19() in test_rcp.cpp.o
      f(Mesh&) in test_rcp.cpp.o
      f_const(Mesh const&) in test_rcp.cpp.o
      ____C_A_T_C_H____T_E_S_T____49() in test_rcp.cpp.o
      ____C_A_T_C_H____T_E_S_T____63() in test_rcp.cpp.o
      f2_const(Mesh2 const&) in test_rcp.cpp.o
      f2_hybrid(Mesh2&) in test_rcp.cpp.o
```

Reply:
What does `/Users/jp/Programs/GitHub/Spack/views/deal_II_dev/bin/ar` point to?

Reply:
Also can you set `LDFLAGS=-Wl,-arch,x86_64`, remove the build directory and start again?

Reply:
Thanks for your patience @isuruf. The output is here
```
$ /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/bin/ar -V
GNU ar (GNU Binutils) 2.28
Copyright (C) 2017 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) any later version.
This program has absolutely no warranty.
```
This version of binutils was also built through Spack
```
$ readlink /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/bin/ar
/Users/jp/Programs/GitHub/Spack/spack/opt/spack/darwin-elcapitan-x86_64/clang-8.0.0-apple/binutils-2.28-qn4eamd42wvsbgwuqadkb4v4vwl4oete/bin/ar
```
and also
```
$ /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/bin/mpicxx -v
Apple LLVM version 8.0.0 (clang-800.0.42.1)
Target: x86_64-apple-darwin15.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```
so I would have thought that this should constitute part of a consistent toolchain. But if you don't think that this is the case then we could just ignore the issue altogether.

To response to the second suggestion, I picked up these flags another way. I make the following change
```
# If SymEngine is built with shared libs on then the static linker
# settings are empty. This duplicates the settings already correctly
# preset for linking in the main shared library.
MESSAGE("old CMAKE_STATIC_LINKER_FLAGS: ${CMAKE_STATIC_LINKER_FLAGS}")
if(BUILD_SHARED_LIBS)
  set(CMAKE_STATIC_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS}")
endif()
MESSAGE("new CMAKE_STATIC_LINKER_FLAGS: ${CMAKE_STATIC_LINKER_FLAGS}")

add_library(catch STATIC ${SRC})
# target_link_libraries(catch symengine)
```  
with the output
```
old CMAKE_STATIC_LINKER_FLAGS:
new CMAKE_STATIC_LINKER_FLAGS: -Wl,-arch,x86_64
```
But this also didn't work due to the use of some invalid syntax
```
[  6%] Linking CXX static library libcatch.a
/Users/jp/Programs/GitHub/Spack/views/deal_II_dev/bin/ar: invalid option -- W
Usage: /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/bin/ar [emulation options] [-]{dmpqrstx}[abcDfilMNoPsSTuvV] [--plugin <name>] [member-name] [count] archive-file file...
       /Users/jp/Programs/GitHub/Spack/views/deal_II_dev/bin/ar -M [<mri-script]
```
When clang is the compiler, does SymEngine expect some other linker to be used?

For me its not essential that I can run the tests locally. Although I would like to add possibility to the run the tests through the spack package at some point in the future, the persistence of this issue would likely prevent that!.

Reply:
Can you try with the archiver `ar` from system instead of spack's one?

Reply:
I configured with `-DCMAKE_AR="/usr/bin/ar" \`, but sorry that didn't work either :-/